### ISSUE TITLE
Does not compile on x86/Ubuntu

### ISSUE THREAD

**Issue by Fiono11** [sfackler/rust-openssl#1448]
```
The following warnings were emitted during compilation:

warning: build/expando.c:4:24: error: pasting "RUST_VERSION_OPENSSL_" and "(" does not give a valid preprocessing token
warning:  #define VERSION2(n, v) RUST_VERSION_##n##_##v
warning:                         ^
warning: build/expando.c:5:23: note: in expansion of macro ‘VERSION2’
warning:  #define VERSION(n, v) VERSION2(n, v)
warning:                        ^~~~~~~~
warning: build/expando.c:10:1: note: in expansion of macro ‘VERSION’
warning:  VERSION(OPENSSL, OPENSSL_VERSION_NUMBER)
warning:  ^~~~~~~

error: failed to run custom build command for `openssl-sys v0.9.61 (/home/fiono/Downloads/rust-openssl/openssl-sys)`

Caused by:
  process didn't exit successfully: `/home/fiono/Downloads/rust-openssl/target/debug/build/openssl-sys-c3ad4e21de3b2ef0/build-script-main` (exit code: 101)
  --- stdout
  cargo:rustc-cfg=const_fn
  cargo:rerun-if-env-changed=X86_64_UNKNOWN_LINUX_GNU_OPENSSL_LIB_DIR
  X86_64_UNKNOWN_LINUX_GNU_OPENSSL_LIB_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_LIB_DIR
  OPENSSL_LIB_DIR unset
  cargo:rerun-if-env-changed=X86_64_UNKNOWN_LINUX_GNU_OPENSSL_INCLUDE_DIR
  X86_64_UNKNOWN_LINUX_GNU_OPENSSL_INCLUDE_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_INCLUDE_DIR
  OPENSSL_INCLUDE_DIR unset
  cargo:rerun-if-env-changed=X86_64_UNKNOWN_LINUX_GNU_OPENSSL_DIR
  X86_64_UNKNOWN_LINUX_GNU_OPENSSL_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_DIR
  OPENSSL_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=OPENSSL_STATIC
  cargo:rerun-if-env-changed=OPENSSL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_STATIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=SYSROOT
  cargo:rerun-if-env-changed=OPENSSL_STATIC
  cargo:rerun-if-env-changed=OPENSSL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_STATIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_DYNAMIC
  cargo:rustc-link-search=native=/usr/local/lib
  cargo:rustc-link-lib=ssl
  cargo:rustc-link-lib=crypto
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=OPENSSL_STATIC
  cargo:rerun-if-env-changed=OPENSSL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_STATIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  OPT_LEVEL = Some("0")
  TARGET = Some("x86_64-unknown-linux-gnu")
  HOST = Some("x86_64-unknown-linux-gnu")
  CC_x86_64-unknown-linux-gnu = None
  CC_x86_64_unknown_linux_gnu = None
  HOST_CC = None
  CC = None
  CFLAGS_x86_64-unknown-linux-gnu = None
  CFLAGS_x86_64_unknown_linux_gnu = None
  HOST_CFLAGS = None
  CFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("true")
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")
  running: "cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-I" "/usr/local/include" "-Wall" "-Wextra" "-E" "build/expando.c"
  cargo:warning=build/expando.c:4:24: error: pasting "RUST_VERSION_OPENSSL_" and "(" does not give a valid preprocessing token
  cargo:warning= #define VERSION2(n, v) RUST_VERSION_##n##_##v
  cargo:warning=                        ^
  cargo:warning=build/expando.c:5:23: note: in expansion of macro ‘VERSION2’
  cargo:warning= #define VERSION(n, v) VERSION2(n, v)
  cargo:warning=                       ^~~~~~~~
  cargo:warning=build/expando.c:10:1: note: in expansion of macro ‘VERSION’
  cargo:warning= VERSION(OPENSSL, OPENSSL_VERSION_NUMBER)
  cargo:warning= ^~~~~~~
  exit code: 1

  --- stderr
  thread 'main' panicked at '
  Header expansion error:
  Error { kind: ToolExecError, message: "Command \"cc\" \"-O0\" \"-ffunction-sections\" \"-fdata-sections\" \"-fPIC\" \"-g\" \"-fno-omit-frame-pointer\" \"-m64\" \"-I\" \"/usr/local/include\" \"-Wall\" \"-Wextra\" \"-E\" \"build/expando.c\" with args \"cc\" did not execute successfully (status code exit code: 1)." }

  Failed to find OpenSSL development headers.

  You can try fixing this setting the `OPENSSL_DIR` environment variable
  pointing to your OpenSSL installation or installing OpenSSL headers package
  specific to your distribution:

      # On Ubuntu
      sudo apt-get install libssl-dev
      # On Arch Linux
      sudo pacman -S openssl
      # On Fedora
      sudo dnf install openssl-devel

  See rust-openssl README for more information:

      https://github.com/sfackler/rust-openssl#linux
  ', openssl-sys/build/main.rs:147:13
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
error: build failed
```

Already tried `sudo apt-get install libssl-dev` but still doesn't work.

**Comment by sfackler** [comment]
Do you have the OpenSSL 3.0.0 alpha installed?

**Comment by Fiono11** [comment]
Yes.

**Comment by sfackler** [comment]
The released version of this library does not support 3.0.0. You can use the `ossl-300` branch off of this repository until OpenSSL cuts a stable 3.0.0 release.

**Comment by Fiono11** [comment]
I didn't have OpenSSL 3.0.0 alpha installed, installed it just now because I thought it would resolve the problem ahah. So that means that wasn't the issue...The problem is I am using other rust libraries that compile from `main` branch. Is there other way I can resolve the issue?

**Comment by sfackler** [comment]
You can use a stable release of OpenSSL.

**Comment by sfackler** [comment]
Oh wait, but you're saying that you didn't have that installed initially? What are the contents of your `openssl/opensslv.h` header then?

**Comment by Fiono11** [comment]
```  GNU nano 2.9.3                      opensslv.h                                

 */

/*
 * Shared library version
 *
 * This is strictly to express ABI version, which may or may not
 * be related to the API version expressed with the macros above.
 * This is defined in free form.
 */
# define OPENSSL_SHLIB_VERSION 3

/*
 * SECTION 2: USEFUL MACROS
 */

/* For checking general API compatibility when preprocessing */
# define OPENSSL_VERSION_PREREQ(maj,min)                                \
    ((OPENSSL_VERSION_MAJOR << 16) + OPENSSL_VERSION_MINOR >= ((maj) << 16) + ($


^G Get Help  ^O Write Out ^W Where Is  ^K Cut Text  ^J Justify   ^C Cur Pos
^X Exit      ^R Read File ^\ Replace   ^U Uncut Text^T To Spell  ^_ Go To Line
```

**Comment by sfackler** [comment]
That header is from an OpenSSL 3.0.0 alpha 1 install.

**Comment by ghost-in-the-zsh** [comment]
I've run into this issue, same kind of error, but I'm using what seems to be an actual stable release of OpenSSL:
```
$ openssl version
OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022)
```

The relevant section of my `/usr/include/openssl/opensslv.h` file has this:
```
 25 /*
 26  * Base version macros
 27  *
 28  * These macros express version number MAJOR.MINOR.PATCH exactly
 29  */
 30 # define OPENSSL_VERSION_MAJOR  3
 31 # define OPENSSL_VERSION_MINOR  0
 32 # define OPENSSL_VERSION_PATCH  2
 33 
 34 /*
 35  * Additional version information
 36  *
 37  * These are also part of the new version scheme, but aren't part
 38  * of the version number itself.
 39  */
 40 
 41 /* Could be: #define OPENSSL_VERSION_PRE_RELEASE "-alpha.1" */
 42 # define OPENSSL_VERSION_PRE_RELEASE ""
 43 /* Could be: #define OPENSSL_VERSION_BUILD_METADATA "+fips" */
 44 /* Could be: #define OPENSSL_VERSION_BUILD_METADATA "+vendor.1" */
 45 # define OPENSSL_VERSION_BUILD_METADATA ""
```
How can I address this issue when trying to `cargo build` an app?

**Comment by ghost-in-the-zsh** [comment]
I was able to solve the problem. I had to make this change to my `Cargo.toml` file:

```diff
- reqwest = { version = "0.11", features = ["blocking", "json"] }
+ reqwest = { version = "0.11", default-features = false, features = ["blocking", "json", "rustls-tls-native-roots"] }
```

The odd thing is that there's another laptop that's also running Ubuntu 22.04 LTS where this was *not* needed. The main difference between the laptops is that the one that ran into the above issue is using a *fresh install* whereas the one *without* the issue was simply *upgraded* to 22.04 LTS from previous versions (been using it since 18.04).

**Comment by ghost-in-the-zsh** [comment]
I was able to solve the problem. I had to make this change to my `Cargo.toml` file:

```diff
- reqwest = { version = "0.11", features = ["blocking", "json"] }
+ reqwest = { version = "0.11", default-features = false, features = ["blocking", "json", "rustls-tls-native-roots"] }
```

The odd thing is that there's another laptop that's also running Ubuntu 22.04 LTS where this was *not* needed. The main difference between the laptops is that the one that ran into the above issue is using a *fresh install* whereas the one *without* the issue was simply *upgraded* to 22.04 LTS from previous versions (been using it since 18.04).
